{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Discrete morphology - Tree Inference\n",
    "## Phylogenetic inference with discrete morphological data\n",
    "### April M. Wright, Michael J. Landis and Sebastian Höhna\n",
    "\n",
    "prerequisites:\n",
    "- [Intro](https://revbayes.github.io/tutorials/intro_graph_models/)\n",
    "- [Intro to Rev Language](https://revbayes.github.io/tutorials/intro/)\n",
    "- [MCMC Archery](https://revbayes.github.io/tutorials/mcmc_archery/)\n",
    "- [CTMC](https://revbayes.github.io/tutorials/ctmc/)\n",
    "\n",
    "\n",
    "\n",
    "# Introduction\n",
    "\n",
    "While molecular data have become the default for building phylogenetic\n",
    "trees for many types of evolutionary analysis, morphological data\n",
    "remains important, particularly for analyses involving fossils. The use\n",
    "of morphological data raises special considerations for model-based\n",
    "methods for phylogenetic inference. Morphological data are typically\n",
    "collected to maximize the number of parsimony-informative characters -\n",
    "that is, the characters that provide information in favor of one\n",
    "topology over another. Morphological characters also do not carry common\n",
    "meanings from one character in a matrix to the next; character codings\n",
    "are made arbitrarily. These two factors require extensions to our\n",
    "existing phylogenetic models. Accounting for the complexity of\n",
    "morphological characters remains challenging. This tutorial will provide\n",
    "a discussion of modeling morphological characters, and will demonstrate\n",
    "how to perform Bayesian phylogenetic analysis with morphology using\n",
    "RevBayes \\cite{Hoehna2016b}.\n",
    "\n",
    "\n",
    "\n",
    "## Overview of Discrete Morphology Models %}\n",
    "\n",
    "<img src=\"figures/tikz/Mk_model.png\" /> \n",
    "_Graphical model showing the Mk model (left panel)._ \n",
    "_Rev code specifying the Mk model is on the right-hand panel._\n",
    "\n",
    "Molecular data forms the basis of most phylogenetic analyses today.\n",
    "However, morphological characters remain relevant: Fossils often provide\n",
    "our only direct observation of extinct biodiversity; DNA degradation can\n",
    "make it difficult or impossible to obtain sufficient molecular data from\n",
    "fragile museum specimens. Using morphological data can help researchers\n",
    "include specimens in their phylogeny that might be left out of a\n",
    "molecular tree.\n",
    "\n",
    "To understand how morphological characters are modeled, it is important\n",
    "to understand how characters are collected. Unlike in molecular data,\n",
    "for which homology is algorithmically determined, homology in a\n",
    "character is typically assessed by an expert. Biologists will typically\n",
    "decide what characters are homologous by looking across specimens at the\n",
    "same structure in multiple taxa; they may also look at the developmental\n",
    "origin of structures in making this assessment \\cite{Phillips06}. Once\n",
    "homology is determined, characters are broken down into states, or\n",
    "different forms a single character can take. The state `0` commonly\n",
    "refers to absence, meaning that character is not present. In some\n",
    "codings, absence will mean that character has not evolved in that group.\n",
    "In others, absence means that that character has not evolved in that\n",
    "group, and/or that that character has been lost in that group\n",
    "\\cite{freudenstein05}. This type of coding is arbitrary, but both\n",
    "*non-random* and *meaningful*, and poses challenges for how we model\n",
    "the data.\n",
    "\n",
    "Historically, most phylogenetic analyses using morphological characters\n",
    "have been performed using the maximum parsimony optimality criterion.\n",
    "Maximum parsimony analysis involves proposing trees from the\n",
    "morphological data. Each tree is evaluated according to how many changes\n",
    "it implied in the data, and the tree that requires the fewest changes is\n",
    "preferred. In this way of estimating a tree, a character that does not\n",
    "change, or changes only in one taxon, cannot be used to discriminate\n",
    "between trees (i.e., it does not favor a topology). Therefore, workers\n",
    "with parsimony typically do not collect characters that are parsimony\n",
    "uninformative.\n",
    "\n",
    "In 2001, Paul Lewis \\cite{Lewis2001} introduced a generalization of the\n",
    "Jukes-Cantor model of sequence evolution for use with morphological\n",
    "data. This model, called the Mk (Markov model, assuming each character\n",
    "is in one of *k* states) model provided a mathematical formulation that\n",
    "could be used to estimate trees from morphological data in both\n",
    "likelihood and Bayesian frameworks. While this model is a useful step\n",
    "forward, as a generalization of the Jukes-Cantor, it still makes fairly\n",
    "simplistic assumptions. This tutorial will guide you through estimating\n",
    "a phylogeny with the Mk model, and two useful extensions to the model.\n",
    "\n",
    "\n",
    "## The Mk Model \n",
    "\n",
    "The Mk model is a generalization of the Jukes-Cantor model of nucleotide\n",
    "sequence evolution, which we discussed in the [CTMC](https://revbayes.github.io/tutorials/ctmc/) tutorial. \n",
    "The Q matrix for a two-state Mk model looks like so:\n",
    "\n",
    "$$Q = \\begin{pmatrix} -\\mu_0 & \\mu_{01} \\\\\n",
    "\\mu_{10} & -\\mu_1  &\\\\\n",
    "\\end{pmatrix} \\mbox{  ,}$$\n",
    "\n",
    "This matrix can be expanded to accommodate multi-state data, as well:\n",
    "\n",
    "$$Q = \\begin{pmatrix} -\\mu_0 & \\mu_{01} & \\mu_{02} & \\mu_{03} \\\\\n",
    "\\mu_{10} & -\\mu_1  & \\mu_{12} & \\mu_{13} \\\\\n",
    "\\mu_{20} & \\mu_{21} & -\\mu_2  & \\mu_{23} \\\\\n",
    "\\mu_{30} & \\mu_{31} & \\mu_{32} & -\\mu_3 \n",
    "\\end{pmatrix} \\mbox{  ,}$$\n",
    "\n",
    "However, the Mk model sets transitions to be equal from any state to any\n",
    "other state. In that sense, our multistate matrix really looks like\n",
    "this:\n",
    "\n",
    "$$Q = \\begin{pmatrix} -(k-1)\\mu & \\mu & \\mu & \\mu \\\\\n",
    "\\mu & -(k-1)\\mu  & \\mu & \\mu \\\\\n",
    "\\mu & \\mu & -(k-1)\\mu  & \\mu \\\\\n",
    "\\mu & \\mu & \\mu & -(k-1)\\mu \\\\\n",
    "\\end{pmatrix} \\mbox{  ,}$$\n",
    "\n",
    "Because this is a Jukes-Cantor-like model \\cite{Jukes1969}, state\n",
    "frequencies do not vary as a model parameter. These assumptions may seem\n",
    "unrealistic. However, all models are a compromise between reality and\n",
    "generalizability. Prior work has demonstrated that, in many conditions,\n",
    "the model does perform adequately \\cite{Wright2014}. Because morphological\n",
    "characters do not carry common meaning across sites in a matrix in the\n",
    "way that nucleotide characters do, making assumptions that fit all\n",
    "characters is challenging. A visualization of this simple model can be\n",
    "seen in above graphical model.\n",
    "\n",
    "We will first perform a phylogenetic analysis using the Mk model. In\n",
    "further sections, we will explore how to relax key assumptions of the Mk\n",
    "model.\n",
    "\n",
    "## Ascertainment Bias \n",
    "\n",
    "When Lewis first introduced the Mk model, he observed that branch\n",
    "lengths on the trees were greatly inflated. The reason for this is that\n",
    "when morphological characters are collected, characters that do not\n",
    "vary, or vary in a non-parsimony-informative way (such as\n",
    "autapomorphies) are excluded. Excluding these low-rate characters causes\n",
    "the overall amount of evolution to be over-estimated. This causes an\n",
    "inflation in the branch lengths \\cite{Lewis2001}.\n",
    "\n",
    "Therefore, when performing a morphological phylogenetic analysis, it is\n",
    "important to correct for this bias. There are numerous statistically\n",
    "valid ways to perform this correction \\cite{allman08b}. Original corrections\n",
    "simulated invariant and non-parsimony informative characters along the\n",
    "proposed tree. The likelihood of these characters would then be\n",
    "calculated and used to normalize the total likelihood value. RevBayes\n",
    "implements a dynamic programming approach that calculates the same\n",
    "likelihood, but does so faster.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "### Example: Inferring a Phylogeny of Fossil Bears Using the Mk Model \n",
    "\n",
    "In this example, we will use morphological character data from 18 taxa\n",
    "of extinct bears \\cite{abella2011}. The dataset contains 62 binary\n",
    "characters, a fairly typical dataset size for morphological characters.\n",
    "\n",
    "\n",
    "\n",
    "**Tutorial Format**\n",
    "\n",
    "This tutorial follows a specific format for issuing instructions and\n",
    "information.\n",
    "\n",
    ">The boxed instructions guide you to complete tasks that are not part of\n",
    ">the RevBayes syntax, but rather direct you to create directories or\n",
    ">files or similar.\n",
    "\n",
    "Information describing the commands and instructions will be written in\n",
    "paragraph-form before or after they are issued.\n",
    "\n",
    "All command-line text, including all Rev syntax, are given in\n",
    "`monotype font`. Furthermore, blocks of Rev code that are needed to\n",
    "build the model, specify the analysis, or execute the run are given in\n",
    "separate shaded boxes. For example, we will instruct you to create a\n",
    "constant node called `example` that is equal to `1.0` using the `<-`\n",
    "operator like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "    example <- 1.0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It is important to be aware that some PDF viewers may render some\n",
    "characters given as differently. Thus, if you copy and paste text from\n",
    "this PDF, you may introduce some incorrect characters. Because of this,\n",
    "we recommend that you type the instructions in this tutorial or copy\n",
    "them from the scripts provided.\n",
    "\n",
    "\n",
    "**Data and Files**\n",
    "\n",
    ">On your own computer, create a directory called **data**.\n",
    ">In this directory download the data\n",
    ">files: [`bears.nex`](data/bears.nex).\n",
    "\n",
    "\n",
    "\n",
    "**Getting Started**\n",
    "\n",
    ">Create a new directory (in `RB_DiscreteMorphology_Tutorial`) called **scripts**.\n",
    ">(If you do not have this folder, please refer to the directions in\n",
    ">section `Data and Files`.\n",
    "\n",
    "When you execute RevBayes in this exercise, you will do so within the\n",
    "main directory you created (`RB_DiscreteMorphology_Tutorial`), thus,\n",
    "if you are using a Unix-based operating system, we recommend that you\n",
    "add the RevBayes binary to your path. Alternatively make sure that you set the \n",
    "working directory to, for example, **RB_DiscreteMorphology_Tutorial** if this is \n",
    "the directory you stored the scripts and data in.\n",
    "\n",
    "\n",
    "\n",
    "**Creating Rev Files**\n",
    "\n",
    "In this exercise, you will work primarily in your text editor and\n",
    "create a set of files that will be easily managed and interchanged. \n",
    "In this first section, you will write the following file\n",
    "from scratch and save them in the `scripts` directory:\n",
    "\n",
    "-   `mcmc_mk.Rev`: the *Rev* file that loads the data, \n",
    "specifies the model describing discrete morphological\n",
    "character change (binary characters), \n",
    "and specifies the monitors and MCMC sampler.\n",
    "\n",
    "All of the files that you will create are also provided in the\n",
    "RevBayes tutorial here (see the top of this webpage). \n",
    "Please refer to these files to verify or troubleshoot your own scripts.\n",
    "\n",
    ">Open your text editor and create the Rev-script file called **mcmc_Mk.Rev** in the\n",
    ">`scripts` directory.\n",
    ">\n",
    ">Enter the Rev code provided in this section in the new file.\n",
    "\n",
    "\n",
    "In this section you will begin the file and write the Rev commands for\n",
    "loading in the taxon list and managing the data matrices. Then, starting\n",
    "in section Mk Model, you will move on to specifying each of \n",
    "the model components. Once the model specifications are\n",
    "complete, you will complete the script with the instructions given in section\n",
    "**subsec_complete_MCMC**.\n",
    "\n",
    "\n",
    "**Load Data Matrices**\n",
    "\n",
    "RevBayes uses the function `readDiscreteCharacterData()` to load a\n",
    "data matrix to the workspace from a formatted file. This function can be\n",
    "used for both molecular sequences and discrete morphological characters.\n",
    "Import the morphological character matrix and assign it to the variable\n",
    "`morpho`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   Successfully read one character matrix from file 'data/bears.nex'\n"
     ]
    }
   ],
   "source": [
    "    morpho <- readDiscreteCharacterData(\"data/bears.nex\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Create Helper Variables**\n",
    "\n",
    "Before we begin writing the Rev scripts for each of the models, \n",
    "we need to instantiate a couple “helper variables” that will\n",
    "be used by downstream parts of our model specification. \n",
    "\n",
    "Create a new constant node called `num_taxa` that is equal to the number\n",
    "of species in our analysis (18) and a constant node called `num_branches` representing\n",
    "the number of branches in the tree. We will also create a constant node of\n",
    "the taxon names. This list will be used to initialize the tree."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "    taxa <- morpho.names()\n",
    "    num_taxa <- morpho.size() \n",
    "    num_branches <- 2 * num_taxa - 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, create two workspace variables called `mvi` and `mni`. These variable are \n",
    "iterators that will build a vector containing all of the MCMC moves used\n",
    "to propose new states for every stochastic node in the model graph. Each\n",
    "time a new move is added to the vector, `mvi` will be incremented by a\n",
    "value of `1`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "    mvi = 1\n",
    "    mni = 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One important distinction here is that `mvi` is part of the RevBayes\n",
    "workspace and not the hierarchical model. Thus, we use the workspace\n",
    "assignment operator `=` instead of the constant node assignment `<-`.\n",
    "\n",
    "\n",
    "## The Mk Model \n",
    "\n",
    "First, we will create a joint prior on the branch lengths and tree topology. \n",
    "This should be familiar from the {% page_ref ctmc %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "    br_len_lambda ~ dnExp(0.2)\n",
    "    moves[mvi++] = mvScale(br_len_lambda, weight=2)\n",
    "\n",
    "    phylogeny ~ dnUniformTopologyBranchLength(taxa, branchLengthDistribution=dnExponential(br_len_lambda))\n",
    "    moves[mvi++] = mvNNI(phylogeny, weight=num_branches/2.0)\n",
    "    moves[mvi++] = mvSPR(phylogeny, weight=num_branches/10.0)\n",
    "    moves[mvi++] = mvBranchLengthScale(phylogeny, weight=num_branches)\n",
    "    \n",
    "    tree_length := phylogeny.treeLength()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we will create a $Q$ matrix. Recall that the Mk model is simply a\n",
    "generalization of the JC model. Therefore, we will create a 2x2 $Q$ matrix\n",
    "using `fnJC`, which initializes Q-matrices with equal transition\n",
    "probabilities between all states."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "    Q_morpho <- fnJC(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have the basics of the model specified, we will add\n",
    "Gamma-distributed rate variation and specify moves on the parameter to\n",
    "the Gamma distribution."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "    alpha_morpho ~ dnUniform( 0, 1E6 )\n",
    "    rates_morpho := fnDiscretizeGamma( alpha_morpho, alpha_morpho, 4 )\n",
    "\n",
    "    #Moves on the parameters to the Gamma distribution.\n",
    "    moves[mvi++] = mvScale(alpha_morpho, lambda=1, weight=2.0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Lastly, we set up the CTMC. This should also be familiar from the {% page_ref ctmc %}. \n",
    "We see some familiar pieces: tree, $Q$ matrix and site_rates.\n",
    "We also have two new keywords: data type and coding. The data type\n",
    "argument specifies the type of data - in our case, “Standard”, the\n",
    "specification for morphology."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "    phyMorpho ~ dnPhyloCTMC(tree=phylogeny, siteRates=rates_morpho, Q=Q_morpho, type=\"Standard\")\n",
    "    phyMorpho.clamp(morpho)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All of the components of the model are now specified.\n",
    "\n",
    "\n",
    "\n",
    "### Complete MCMC Analysis\n",
    "\n",
    "**Create Model Object**\n",
    "\n",
    "We can now create our workspace model variable with our fully specified\n",
    "model DAG. We will do this with the `model()` function and provide a\n",
    "single node in the graph (`phylogeny`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "    mymodel = model(phylogeny)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The object `mymodel` is a wrapper around the entire model graph and\n",
    "allows us to pass the model to various functions that are specific to\n",
    "our MCMC analysis.\n",
    "\n",
    "\n",
    "**Specify Monitors and Output Filenames**\n",
    "\n",
    "The next important step for our Rev-script is to specify the\n",
    "monitors and output file names. For this, we create a vector called\n",
    "`monitors` that will each sample and record or output our MCMC.\n",
    "\n",
    "The first monitor we will create will monitor every named random\n",
    "variable in our model graph. This will include every stochastic and\n",
    "deterministic node using the `mnModel` monitor. The only parameter that\n",
    "is not included in the `mnModel` is the tree topology. Therefore, the\n",
    "parameters in the file written by this monitor are all numerical\n",
    "parameters written to a tab-separated text file that can be opened by\n",
    "accessory programs for evaluating such parameters. We will also name the\n",
    "output file for this monitor and indicate that we wish to sample our\n",
    "MCMC every 10 cycles."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "    monitors[mni++] = mnModel(filename=\"output/mk.log\", printgen=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `mnFile` monitor writes any parameter we specify to file. Thus, if\n",
    "we only cared about the branch lengths and nothing else (this is not a\n",
    "typical or recommended attitude for an analysis this complex) we\n",
    "wouldn't use the `mnModel` monitor above and just use the `mnFile`\n",
    "monitor to write a smaller and simpler output file. Since the tree\n",
    "topology is not included in the `mnModel` monitor (because it is not\n",
    "numerical), we will use `mnFile` to write the tree to file by specifying\n",
    "our `phylogeny` variable in the arguments."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "    monitors[mni++] = mnFile(filename=\"output/mk.trees\", printgen=10, phylogeny)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The third monitor we will add to our analysis will print information to\n",
    "the screen. Like with `mnFile` we must tell `mnScreen` which parameters\n",
    "we'd like to see updated on the screen."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "    monitors[mni++] = mnScreen(printgen=100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set-Up the MCMC\n",
    "\n",
    "Once we have set up our model, moves, and monitors, we can now create\n",
    "the workspace variable that defines our MCMC run. We do this using the\n",
    "`mcmc()` function that simply takes the three main analysis components\n",
    "as arguments."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "    mymcmc = mcmc(mymodel, monitors, moves, nruns=2, combine=\"mixed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The MCMC object that we named `mymcmc` has a member method called\n",
    "`.run()`. This will execute our analysis and we will set the chain\n",
    "length to `20000` cycles using the `generations` option."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "   Running MCMC simulation\n",
      "   This simulation runs 2 independent replicates.\n",
      "   The simulator uses 9 different moves in a random move schedule with 82.8 moves per iteration\n",
      "\n",
      "Iter        |      Posterior   |     Likelihood   |          Prior   |    elapsed   |        ETA   |\n",
      "----------------------------------------------------------------------------------------------------\n",
      "0           |       -845.264   |       -641.161   |       -204.103   |   00:00:00   |   --:--:--   |\n",
      "100         |       -834.833   |       -641.161   |       -193.672   |   00:00:01   |   --:--:--   |\n",
      "200         |        -840.93   |       -641.161   |       -199.769   |   00:00:01   |   00:01:39   |\n",
      "300         |       -869.307   |       -641.161   |       -228.146   |   00:00:02   |   00:02:11   |\n",
      "400         |       -854.517   |       -641.161   |       -213.356   |   00:00:02   |   00:01:38   |\n",
      "500         |       -865.815   |       -641.161   |       -224.654   |   00:00:03   |   00:01:57   |\n",
      "600         |       -858.304   |       -641.161   |       -217.143   |   00:00:04   |   00:02:09   |\n",
      "700         |       -464.404   |       -436.021   |       -28.3833   |   00:00:04   |   00:01:50   |\n",
      "800         |       -440.868   |       -417.594   |       -23.2741   |   00:00:05   |   00:02:00   |\n",
      "900         |       -430.017   |       -413.235   |        -16.782   |   00:00:05   |   00:01:46   |\n",
      "1000        |       -435.553   |           -420   |       -15.5526   |   00:00:06   |   00:01:54   |\n",
      "1100        |        -430.32   |       -416.379   |       -13.9413   |   00:00:06   |   00:01:43   |\n",
      "1200        |       -430.092   |       -417.336   |       -12.7563   |   00:00:07   |   00:01:49   |\n",
      "1300        |       -444.311   |       -427.008   |       -17.3037   |   00:00:08   |   00:01:55   |\n",
      "1400        |       -424.509   |       -415.913   |       -8.59553   |   00:00:08   |   00:01:46   |\n",
      "1500        |        -435.63   |       -424.284   |       -11.3463   |   00:00:09   |   00:01:51   |\n",
      "1600        |       -427.009   |       -417.695   |       -9.31368   |   00:00:10   |   00:01:55   |\n",
      "1700        |       -434.242   |       -420.407   |       -13.8349   |   00:00:10   |   00:01:47   |\n",
      "1800        |       -436.687   |       -416.622   |       -20.0643   |   00:00:11   |   00:01:51   |\n",
      "1900        |       -423.646   |       -415.863   |       -7.78282   |   00:00:11   |   00:01:44   |\n",
      "\n",
      "\n",
      "Iter        |      Posterior   |     Likelihood   |          Prior   |    elapsed   |        ETA   |\n",
      "----------------------------------------------------------------------------------------------------\n",
      "2000        |       -433.295   |       -421.026   |       -12.2682   |   00:00:12   |   00:01:48   |\n",
      "2100        |       -441.501   |       -424.961   |       -16.5405   |   00:00:12   |   00:01:42   |\n",
      "2200        |       -440.657   |       -422.187   |         -18.47   |   00:00:13   |   00:01:45   |\n",
      "2300        |       -432.152   |       -418.696   |       -13.4558   |   00:00:14   |   00:01:47   |\n",
      "2400        |       -428.499   |       -413.981   |       -14.5177   |   00:00:14   |   00:01:42   |\n",
      "2500        |       -438.617   |       -418.824   |       -19.7935   |   00:00:15   |   00:01:45   |\n",
      "2600        |       -438.052   |       -420.055   |       -17.9968   |   00:00:15   |   00:01:40   |\n",
      "2700        |       -431.422   |       -418.244   |       -13.1788   |   00:00:16   |   00:01:42   |\n",
      "2800        |       -437.758   |       -423.311   |        -14.447   |   00:00:16   |   00:01:38   |\n",
      "2900        |       -430.101   |       -413.909   |       -16.1919   |   00:00:17   |   00:01:40   |\n",
      "3000        |       -440.733   |       -420.677   |       -20.0559   |   00:00:17   |   00:01:36   |\n",
      "3100        |       -429.374   |       -414.794   |       -14.5805   |   00:00:18   |   00:01:38   |\n",
      "3200        |       -426.266   |       -411.116   |       -15.1493   |   00:00:19   |   00:01:39   |\n",
      "3300        |       -441.306   |       -424.265   |        -17.041   |   00:00:19   |   00:01:36   |\n",
      "3400        |       -433.363   |       -419.901   |       -13.4612   |   00:00:20   |   00:01:37   |\n",
      "3500        |       -443.312   |       -431.219   |       -12.0931   |   00:00:20   |   00:01:34   |\n",
      "3600        |       -440.412   |       -421.341   |       -19.0711   |   00:00:21   |   00:01:35   |\n",
      "3700        |       -441.442   |       -422.901   |       -18.5408   |   00:00:21   |   00:01:32   |\n",
      "3800        |       -438.955   |       -430.559   |       -8.39592   |   00:00:22   |   00:01:33   |\n",
      "3900        |       -432.528   |       -418.103   |       -14.4254   |   00:00:23   |   00:01:34   |\n",
      "\n",
      "\n",
      "Iter        |      Posterior   |     Likelihood   |          Prior   |    elapsed   |        ETA   |\n",
      "----------------------------------------------------------------------------------------------------\n",
      "4000        |       -435.365   |       -417.016   |       -18.3489   |   00:00:23   |   00:01:32   |\n",
      "4100        |       -451.423   |       -427.848   |       -23.5751   |   00:00:24   |   00:01:33   |\n",
      "4200        |       -427.618   |       -420.961   |         -6.657   |   00:00:25   |   00:01:34   |\n",
      "4300        |       -447.077   |       -426.203   |       -20.8735   |   00:00:25   |   00:01:31   |\n",
      "4400        |       -436.772   |       -423.293   |       -13.4788   |   00:00:25   |   00:01:28   |\n",
      "4500        |       -426.914   |        -417.79   |       -9.12399   |   00:00:26   |   00:01:29   |\n",
      "4600        |       -439.007   |       -426.983   |       -12.0235   |   00:00:26   |   00:01:27   |\n",
      "4700        |       -438.262   |       -422.457   |       -15.8049   |   00:00:27   |   00:01:27   |\n",
      "4800        |       -422.671   |       -418.543   |        -4.1282   |   00:00:27   |   00:01:25   |\n",
      "4900        |       -435.154   |       -420.073   |       -15.0814   |   00:00:28   |   00:01:26   |\n",
      "5000        |       -431.654   |       -413.808   |       -17.8455   |   00:00:28   |   00:01:24   |\n",
      "5100        |       -430.401   |       -417.288   |       -13.1137   |   00:00:29   |   00:01:24   |\n",
      "5200        |       -440.379   |       -418.869   |         -21.51   |   00:00:29   |   00:01:22   |\n",
      "5300        |       -428.596   |       -418.107   |       -10.4893   |   00:00:30   |   00:01:23   |\n",
      "5400        |       -436.273   |       -417.276   |       -18.9975   |   00:00:30   |   00:01:21   |\n",
      "5500        |       -428.342   |       -414.207   |       -14.1352   |   00:00:31   |   00:01:21   |\n",
      "5600        |       -435.137   |        -416.65   |       -18.4864   |   00:00:31   |   00:01:19   |\n",
      "5700        |       -428.414   |        -413.05   |       -15.3635   |   00:00:32   |   00:01:20   |\n",
      "5800        |       -432.604   |        -422.55   |       -10.0542   |   00:00:32   |   00:01:18   |\n",
      "5900        |       -432.546   |       -420.564   |       -11.9819   |   00:00:33   |   00:01:18   |\n",
      "\n",
      "\n",
      "Iter        |      Posterior   |     Likelihood   |          Prior   |    elapsed   |        ETA   |\n",
      "----------------------------------------------------------------------------------------------------\n",
      "6000        |       -433.623   |       -419.672   |       -13.9509   |   00:00:33   |   00:01:17   |\n",
      "6100        |       -434.678   |       -420.473   |       -14.2044   |   00:00:34   |   00:01:17   |\n",
      "6200        |       -438.622   |       -423.915   |       -14.7065   |   00:00:35   |   00:01:17   |\n",
      "6300        |       -438.343   |       -415.995   |       -22.3483   |   00:00:35   |   00:01:16   |\n",
      "6400        |       -439.913   |       -423.032   |       -16.8812   |   00:00:36   |   00:01:16   |\n",
      "6500        |       -438.337   |       -421.211   |       -17.1264   |   00:00:36   |   00:01:14   |\n",
      "6600        |       -440.432   |       -425.457   |       -14.9742   |   00:00:37   |   00:01:15   |\n",
      "6700        |       -426.567   |       -420.654   |        -5.9122   |   00:00:37   |   00:01:13   |\n",
      "6800        |        -437.26   |       -423.806   |       -13.4542   |   00:00:38   |   00:01:13   |\n",
      "6900        |       -438.411   |       -423.378   |       -15.0327   |   00:00:39   |   00:01:14   |\n",
      "7000        |       -441.137   |       -424.655   |       -16.4819   |   00:00:39   |   00:01:12   |\n",
      "7100        |       -436.572   |       -420.828   |        -15.744   |   00:00:40   |   00:01:12   |\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7200        |       -442.295   |       -423.091   |       -19.2038   |   00:00:40   |   00:01:11   |\n",
      "7300        |       -435.249   |       -416.212   |       -19.0373   |   00:00:41   |   00:01:11   |\n",
      "7400        |       -436.541   |       -418.841   |       -17.7007   |   00:00:41   |   00:01:09   |\n",
      "7500        |        -441.93   |       -425.084   |        -16.846   |   00:00:42   |   00:01:10   |\n",
      "7600        |       -441.327   |       -420.311   |       -21.0157   |   00:00:43   |   00:01:10   |\n",
      "7700        |       -439.852   |        -422.95   |       -16.9016   |   00:00:43   |   00:01:08   |\n",
      "7800        |       -444.754   |       -426.591   |       -18.1622   |   00:00:44   |   00:01:08   |\n",
      "7900        |       -431.363   |       -415.877   |       -15.4862   |   00:00:44   |   00:01:07   |\n",
      "\n",
      "\n",
      "Iter        |      Posterior   |     Likelihood   |          Prior   |    elapsed   |        ETA   |\n",
      "----------------------------------------------------------------------------------------------------\n",
      "8000        |       -439.737   |       -422.063   |       -17.6737   |   00:00:45   |   00:01:07   |\n",
      "8100        |       -440.912   |         -419.2   |       -21.7117   |   00:00:45   |   00:01:06   |\n",
      "8200        |       -427.779   |       -413.601   |       -14.1788   |   00:00:46   |   00:01:06   |\n",
      "8300        |       -431.393   |       -419.078   |       -12.3156   |   00:00:46   |   00:01:04   |\n",
      "8400        |       -434.807   |       -419.699   |       -15.1075   |   00:00:47   |   00:01:04   |\n",
      "8500        |        -443.08   |        -425.71   |       -17.3703   |   00:00:47   |   00:01:03   |\n",
      "8600        |       -436.787   |       -423.252   |       -13.5348   |   00:00:48   |   00:01:03   |\n",
      "8700        |       -435.898   |       -419.802   |       -16.0969   |   00:00:48   |   00:01:02   |\n",
      "8800        |       -437.393   |       -421.094   |       -16.2991   |   00:00:49   |   00:01:02   |\n",
      "8900        |       -443.424   |       -427.965   |       -15.4598   |   00:00:49   |   00:01:01   |\n",
      "9000        |       -435.277   |       -421.669   |       -13.6088   |   00:00:50   |   00:01:01   |\n",
      "9100        |       -441.084   |       -425.605   |       -15.4794   |   00:00:50   |   00:00:59   |\n",
      "9200        |       -442.084   |       -420.293   |       -21.7906   |   00:00:51   |   00:00:59   |\n",
      "9300        |       -429.019   |       -417.952   |        -11.067   |   00:00:52   |   00:00:59   |\n",
      "9400        |       -432.016   |       -423.722   |        -8.2936   |   00:00:53   |   00:00:59   |\n",
      "9500        |        -444.36   |       -429.083   |        -15.277   |   00:00:54   |   00:00:59   |\n",
      "9600        |       -439.004   |       -420.571   |       -18.4339   |   00:00:54   |   00:00:58   |\n",
      "9700        |       -443.298   |        -430.32   |       -12.9786   |   00:00:55   |   00:00:58   |\n",
      "9800        |       -448.543   |       -429.232   |        -19.311   |   00:00:55   |   00:00:57   |\n",
      "9900        |       -434.027   |       -423.712   |       -10.3151   |   00:00:55   |   00:00:56   |\n",
      "\n",
      "\n",
      "Iter        |      Posterior   |     Likelihood   |          Prior   |    elapsed   |        ETA   |\n",
      "----------------------------------------------------------------------------------------------------\n",
      "10000       |        -427.73   |       -415.013   |       -12.7173   |   00:00:56   |   00:00:56   |\n",
      "10100       |       -424.943   |       -413.085   |       -11.8583   |   00:00:57   |   00:00:55   |\n",
      "10200       |       -435.255   |       -420.179   |       -15.0768   |   00:00:58   |   00:00:55   |\n",
      "10300       |       -429.325   |       -415.287   |       -14.0378   |   00:01:00   |   00:00:56   |\n",
      "10400       |       -424.718   |       -414.314   |       -10.4039   |   00:01:01   |   00:00:56   |\n",
      "10500       |       -442.103   |       -422.497   |       -19.6059   |   00:01:01   |   00:00:55   |\n",
      "10600       |       -437.198   |       -421.486   |       -15.7112   |   00:01:02   |   00:00:54   |\n",
      "10700       |       -425.614   |       -417.633   |        -7.9811   |   00:01:03   |   00:00:54   |\n",
      "10800       |       -433.385   |       -417.939   |       -15.4456   |   00:01:03   |   00:00:53   |\n",
      "10900       |       -432.095   |       -413.715   |       -18.3802   |   00:01:04   |   00:00:53   |\n",
      "11000       |       -435.688   |        -421.06   |       -14.6281   |   00:01:04   |   00:00:52   |\n",
      "11100       |        -428.08   |       -415.138   |       -12.9426   |   00:01:05   |   00:00:52   |\n",
      "11200       |       -430.026   |       -412.809   |        -17.217   |   00:01:06   |   00:00:51   |\n",
      "11300       |       -430.306   |       -419.514   |       -10.7921   |   00:01:06   |   00:00:50   |\n",
      "11400       |       -437.527   |       -425.254   |       -12.2727   |   00:01:07   |   00:00:50   |\n",
      "11500       |       -439.664   |       -424.953   |       -14.7106   |   00:01:07   |   00:00:49   |\n",
      "11600       |       -442.557   |       -428.198   |       -14.3589   |   00:01:08   |   00:00:49   |\n",
      "11700       |       -433.346   |        -418.07   |       -15.2767   |   00:01:08   |   00:00:48   |\n",
      "11800       |       -426.462   |       -416.344   |       -10.1174   |   00:01:09   |   00:00:47   |\n",
      "11900       |       -435.503   |       -419.649   |        -15.854   |   00:01:09   |   00:00:46   |\n",
      "\n",
      "\n",
      "Iter        |      Posterior   |     Likelihood   |          Prior   |    elapsed   |        ETA   |\n",
      "----------------------------------------------------------------------------------------------------\n",
      "12000       |       -439.947   |       -421.104   |       -18.8433   |   00:01:10   |   00:00:46   |\n",
      "12100       |       -434.683   |        -419.96   |       -14.7227   |   00:01:10   |   00:00:45   |\n",
      "12200       |       -431.144   |       -418.446   |       -12.6984   |   00:01:11   |   00:00:45   |\n",
      "12300       |       -450.739   |        -431.17   |       -19.5694   |   00:01:11   |   00:00:44   |\n",
      "12400       |       -433.428   |       -418.639   |        -14.789   |   00:01:12   |   00:00:44   |\n",
      "12500       |       -446.487   |        -427.75   |       -18.7367   |   00:01:13   |   00:00:43   |\n",
      "12600       |       -439.149   |       -419.003   |        -20.146   |   00:01:14   |   00:00:43   |\n",
      "12700       |       -443.979   |       -419.445   |       -24.5338   |   00:01:14   |   00:00:42   |\n",
      "12800       |       -425.175   |        -415.16   |        -10.015   |   00:01:15   |   00:00:42   |\n",
      "12900       |       -438.825   |       -418.318   |       -20.5074   |   00:01:15   |   00:00:41   |\n",
      "13000       |        -431.69   |       -419.714   |       -11.9764   |   00:01:16   |   00:00:40   |\n",
      "13100       |       -431.739   |       -419.631   |       -12.1078   |   00:01:16   |   00:00:40   |\n",
      "13200       |       -435.356   |       -420.517   |       -14.8395   |   00:01:17   |   00:00:39   |\n",
      "13300       |       -433.921   |       -421.151   |       -12.7699   |   00:01:17   |   00:00:38   |\n",
      "13400       |       -437.315   |       -423.341   |       -13.9732   |   00:01:18   |   00:00:38   |\n",
      "13500       |       -429.564   |       -414.686   |       -14.8774   |   00:01:18   |   00:00:37   |\n",
      "13600       |       -446.782   |       -428.312   |       -18.4707   |   00:01:19   |   00:00:37   |\n",
      "13700       |       -434.511   |       -418.154   |       -16.3572   |   00:01:19   |   00:00:36   |\n",
      "13800       |       -437.447   |       -421.665   |       -15.7825   |   00:01:20   |   00:00:35   |\n",
      "13900       |        -445.19   |       -426.843   |       -18.3465   |   00:01:21   |   00:00:35   |\n",
      "\n",
      "\n",
      "Iter        |      Posterior   |     Likelihood   |          Prior   |    elapsed   |        ETA   |\n",
      "----------------------------------------------------------------------------------------------------\n",
      "14000       |       -426.304   |       -413.293   |       -13.0108   |   00:01:22   |   00:00:35   |\n",
      "14100       |       -433.745   |        -419.11   |       -14.6346   |   00:01:22   |   00:00:34   |\n",
      "14200       |       -438.523   |       -423.309   |       -15.2148   |   00:01:23   |   00:00:33   |\n",
      "14300       |       -440.656   |         -422.1   |       -18.5567   |   00:01:24   |   00:00:33   |\n",
      "14400       |       -433.268   |         -423.9   |       -9.36841   |   00:01:24   |   00:00:32   |\n",
      "14500       |       -435.369   |       -418.259   |         -17.11   |   00:01:25   |   00:00:32   |\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14600       |       -432.745   |       -419.823   |        -12.922   |   00:01:26   |   00:00:31   |\n",
      "14700       |       -441.193   |       -426.288   |       -14.9048   |   00:01:26   |   00:00:31   |\n",
      "14800       |       -442.222   |       -426.001   |       -16.2219   |   00:01:27   |   00:00:30   |\n",
      "14900       |       -427.919   |        -420.71   |       -7.20904   |   00:01:28   |   00:00:30   |\n",
      "15000       |       -438.013   |       -422.912   |       -15.1009   |   00:01:28   |   00:00:29   |\n",
      "15100       |       -441.601   |       -422.856   |       -18.7452   |   00:01:29   |   00:00:28   |\n",
      "15200       |        -440.03   |       -423.149   |       -16.8818   |   00:01:30   |   00:00:28   |\n",
      "15300       |       -436.978   |       -423.648   |       -13.3295   |   00:01:31   |   00:00:27   |\n",
      "15400       |       -432.233   |       -419.126   |       -13.1067   |   00:01:31   |   00:00:27   |\n",
      "15500       |       -444.402   |       -430.364   |       -14.0382   |   00:01:32   |   00:00:26   |\n",
      "15600       |       -437.008   |       -425.836   |       -11.1724   |   00:01:32   |   00:00:25   |\n",
      "15700       |       -444.649   |       -428.437   |       -16.2128   |   00:01:33   |   00:00:25   |\n",
      "15800       |       -431.521   |        -420.85   |       -10.6706   |   00:01:34   |   00:00:24   |\n",
      "15900       |       -428.847   |       -414.905   |       -13.9411   |   00:01:34   |   00:00:24   |\n",
      "\n",
      "\n",
      "Iter        |      Posterior   |     Likelihood   |          Prior   |    elapsed   |        ETA   |\n",
      "----------------------------------------------------------------------------------------------------\n",
      "16000       |        -433.07   |       -421.206   |       -11.8637   |   00:01:35   |   00:00:23   |\n",
      "16100       |       -434.482   |       -416.529   |       -17.9524   |   00:01:36   |   00:00:23   |\n",
      "16200       |       -428.567   |       -414.725   |       -13.8422   |   00:01:36   |   00:00:22   |\n",
      "16300       |        -438.85   |       -421.267   |       -17.5825   |   00:01:37   |   00:00:22   |\n",
      "16400       |       -441.403   |       -423.078   |       -18.3254   |   00:01:38   |   00:00:21   |\n",
      "16500       |         -434.7   |       -415.174   |       -19.5263   |   00:01:38   |   00:00:20   |\n",
      "16600       |       -434.575   |       -419.022   |       -15.5524   |   00:01:39   |   00:00:20   |\n",
      "16700       |       -435.687   |       -423.339   |       -12.3478   |   00:01:39   |   00:00:19   |\n",
      "16800       |       -432.387   |       -413.749   |        -18.638   |   00:01:40   |   00:00:19   |\n",
      "16900       |       -434.745   |       -426.249   |       -8.49674   |   00:01:41   |   00:00:18   |\n",
      "17000       |       -428.888   |       -416.626   |       -12.2618   |   00:01:41   |   00:00:17   |\n",
      "17100       |       -436.413   |       -422.892   |       -13.5215   |   00:01:42   |   00:00:17   |\n",
      "17200       |       -432.605   |       -419.234   |       -13.3714   |   00:01:42   |   00:00:16   |\n",
      "17300       |       -445.095   |       -432.515   |         -12.58   |   00:01:43   |   00:00:16   |\n",
      "17400       |       -437.291   |       -422.511   |       -14.7805   |   00:01:43   |   00:00:15   |\n",
      "17500       |       -438.329   |       -424.181   |       -14.1479   |   00:01:44   |   00:00:14   |\n",
      "17600       |       -449.496   |       -429.881   |       -19.6151   |   00:01:45   |   00:00:14   |\n",
      "17700       |       -446.145   |       -423.377   |       -22.7681   |   00:01:45   |   00:00:13   |\n",
      "17800       |       -419.736   |       -411.299   |       -8.43788   |   00:01:46   |   00:00:13   |\n",
      "17900       |       -448.139   |       -430.166   |       -17.9734   |   00:01:47   |   00:00:12   |\n",
      "\n",
      "\n",
      "Iter        |      Posterior   |     Likelihood   |          Prior   |    elapsed   |        ETA   |\n",
      "----------------------------------------------------------------------------------------------------\n",
      "18000       |       -427.675   |       -413.131   |       -14.5438   |   00:01:47   |   00:00:11   |\n",
      "18100       |       -423.125   |       -411.055   |       -12.0702   |   00:01:48   |   00:00:11   |\n",
      "18200       |       -436.824   |       -424.412   |       -12.4125   |   00:01:48   |   00:00:10   |\n",
      "18300       |       -435.851   |       -423.339   |       -12.5122   |   00:01:49   |   00:00:10   |\n",
      "18400       |       -435.872   |       -422.123   |       -13.7484   |   00:01:49   |   00:00:09   |\n",
      "18500       |           -434   |       -418.188   |       -15.8123   |   00:01:50   |   00:00:08   |\n",
      "18600       |       -433.828   |       -424.562   |       -9.26585   |   00:01:51   |   00:00:08   |\n",
      "18700       |       -438.555   |       -419.154   |       -19.4009   |   00:01:51   |   00:00:07   |\n",
      "18800       |        -428.01   |       -413.586   |       -14.4247   |   00:01:52   |   00:00:07   |\n",
      "18900       |       -430.806   |       -418.333   |       -12.4725   |   00:01:53   |   00:00:06   |\n",
      "19000       |       -432.081   |       -418.849   |       -13.2312   |   00:01:53   |   00:00:05   |\n",
      "19100       |       -437.067   |       -423.583   |       -13.4846   |   00:01:54   |   00:00:05   |\n",
      "19200       |       -436.426   |       -421.844   |       -14.5817   |   00:01:54   |   00:00:04   |\n",
      "19300       |       -442.809   |       -428.342   |       -14.4668   |   00:01:55   |   00:00:04   |\n",
      "19400       |       -441.664   |       -423.089   |       -18.5748   |   00:01:56   |   00:00:03   |\n",
      "19500       |       -439.279   |       -429.066   |       -10.2126   |   00:01:57   |   00:00:03   |\n",
      "19600       |       -429.513   |       -419.427   |       -10.0863   |   00:01:58   |   00:00:02   |\n",
      "19700       |       -428.224   |       -416.287   |       -11.9377   |   00:01:58   |   00:00:01   |\n",
      "19800       |       -428.215   |       -413.607   |       -14.6076   |   00:01:59   |   00:00:01   |\n",
      "19900       |       -433.483   |       -419.984   |       -13.4995   |   00:02:00   |   00:00:00   |\n",
      "\n",
      "\n",
      "Iter        |      Posterior   |     Likelihood   |          Prior   |    elapsed   |        ETA   |\n",
      "----------------------------------------------------------------------------------------------------\n",
      "20000       |       -428.617   |       -416.479   |       -12.1372   |   00:02:00   |   00:00:00   |\n"
     ]
    }
   ],
   "source": [
    "    mymcmc.run(generations=20000, tuningInterval=200)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once our Markov chain has terminated, we will want RevBayes to close.\n",
    "Tell the program to quit using the `q()` function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "    q()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## You made it! Save all of your files.\n",
    "\n",
    "{% subsection Execute the MCMC Analysis | subsec_run_MCMC %}\n",
    "\n",
    "With all the parameters specified and all analysis components in place,\n",
    "you are now ready to run your analysis. The Rev scripts you just\n",
    "created will all be used by RevBayes.\n",
    "\n",
    ">Begin by running the RevBayes executable. In Unix systems, type the\n",
    ">following in your terminal (if the RevBayes binary is in your path):\n",
    ">rb\n",
    "\n",
    "Provided that you started RevBayes from the correct directory\n",
    "(`RB_DiscreteMorphology_Tutorial`), you can then use the `source()`\n",
    "function to feed RevBayes your Rev-script file (`mcmc_mk.Rev`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "    source(\"scripts/mcmc_mk.Rev\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This will execute the analysis and you should see the following output\n",
    "(though not the exact same values):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   Syntax error while reading characters \"\"scripts/mcmc_mk.Rev\"\" at position 20 to 40 in command:\n",
      "      Processing file \"scripts/mcmc_mk.Rev\"\n",
      "   Syntax error while reading characters \"one\" at position 22 to 24 in command:\n",
      "      Successfully read one character matrix from file 'data/bears.nex'\n",
      "   Syntax error while reading characters \"simulation\" at position 17 to 26 in command:\n",
      "      Running MCMC simulation\n",
      "   Syntax error while reading characters \"runs\" at position 20 to 23 in command:\n",
      "      This simulation runs 2 independent replicates.\n",
      "   Syntax error while reading characters \"uses\" at position 18 to 21 in command:\n",
      "      The simulator uses 5 different moves in a random move schedule with 58.4 moves per iteration\n",
      "   Syntax error while reading characters \"\u0007\" at position 205 to 206 in command:\n",
      "   Iter        |      Posterior   |     Likelihood   |          Prior   |    elapsed   |        ETA  \n",
      "   |;----------------------------------------------------------------------------------------------------\n",
      "   Syntax error while reading character ':' at position 91 in command:\n",
      "   0           |       -680.054   |       -649.452   |       -30.6022   |   00:00:00   |   --:--:--   |\n",
      "   Syntax error while reading character ':' at position 91 in command:\n",
      "   100         |       -419.885   |       -414.047   |       -5.83883   |   00:00:01   |   --:--:--   |\n",
      "   Syntax error while reading character '\u0000' at position 5455 in command:\n",
      "   200         |       -427.028   |       -417.426   |       -9.60277   |   00:00:01   |   00:00:49   |;300         |       -421.585   |        -417.96   |     \n",
      "     -3.6253   |   00:00:02   |   00:01:04   |;400         |       -431.561   |       -427.124   |       -4.43711   |   00:00:03   |   00:01:12   |;500        \n",
      "   |       -423.507   |       -422.002   |       -1.50428   |   00:00:04   |   00:01:16   |;600         |       -418.061   |       -419.644   |        1.58298  \n",
      "   |   00:00:05   |   00:01:18   |;700         |       -427.552   |       -423.884   |       -3.66793   |   00:00:05   |   00:01:06   |;800         |       \n",
      "   -437.39   |       -424.302   |       -13.0876   |   00:00:06   |   00:01:09   |;900         |       -418.405   |       -413.872   |        -4.5323   |  \n",
      "   00:00:07   |   00:01:10   |;1000        |       -425.641   |       -411.291   |       -14.3491   |   00:00:08   |   00:01:12   |;;...\n"
     ]
    }
   ],
   "source": [
    "   Processing file \"scripts/mcmc_mk.Rev\"\n",
    "   Successfully read one character matrix from file 'data/bears.nex'\n",
    "\n",
    "   Running MCMC simulation\n",
    "   This simulation runs 2 independent replicates.\n",
    "   The simulator uses 5 different moves in a random move schedule with 58.4 moves per iteration\n",
    "\n",
    "Iter        |      Posterior   |     Likelihood   |          Prior   |    elapsed   |        ETA   |\n",
    "----------------------------------------------------------------------------------------------------\n",
    "0           |       -680.054   |       -649.452   |       -30.6022   |   00:00:00   |   --:--:--   |\n",
    "100         |       -419.885   |       -414.047   |       -5.83883   |   00:00:01   |   --:--:--   |\n",
    "200         |       -427.028   |       -417.426   |       -9.60277   |   00:00:01   |   00:00:49   |\n",
    "300         |       -421.585   |        -417.96   |        -3.6253   |   00:00:02   |   00:01:04   |\n",
    "400         |       -431.561   |       -427.124   |       -4.43711   |   00:00:03   |   00:01:12   |\n",
    "500         |       -423.507   |       -422.002   |       -1.50428   |   00:00:04   |   00:01:16   |\n",
    "600         |       -418.061   |       -419.644   |        1.58298   |   00:00:05   |   00:01:18   |\n",
    "700         |       -427.552   |       -423.884   |       -3.66793   |   00:00:05   |   00:01:06   |\n",
    "800         |        -437.39   |       -424.302   |       -13.0876   |   00:00:06   |   00:01:09   |\n",
    "900         |       -418.405   |       -413.872   |        -4.5323   |   00:00:07   |   00:01:10   |\n",
    "1000        |       -425.641   |       -411.291   |       -14.3491   |   00:00:08   |   00:01:12   |\n",
    "\n",
    "..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When the analysis is complete, RevBayes will quit and you will have a\n",
    "new directory called `output` that will contain all of the files you\n",
    "specified with the monitors ({% ref subsubsec_Monitors %}).\n",
    "\n",
    "\n",
    "### Exercises %}\n",
    "\n",
    "1. Run the MCMC analysis in RevBayes.\n",
    "2. Look at the resulting files **mk_run_1.log** and **mk_run_2.log** in Tracer and check for convergence.\n",
    "3. Look at the majority rule consesus tree stored **mk.majrule.tre** and the MAP tree stored in **mk.map.tre** in FigTree.\n",
    "\n",
    "\n",
    "#### Ascertainment Bias\n",
    "\n",
    "As discussed earlier in the section Ascertainment_Bias, we also need to\n",
    "correct for ascertainment bias. \n",
    "\n",
    ">Create a copy of your previous `Rev` script, and call it *mcmc_Mkv.Rev*. \n",
    ">You will need to modify the `Rev`\n",
    ">code provided in this section in this file.\n",
    "\n",
    "In `RevBayes` it is actually very simple to add a correction for ascertainment bias. \n",
    "You only need to set the option `coding=\"variable\"` in the `dnPhyloCTMC`. Coding specifies \n",
    "what type of ascertainment bias is expected. We are using the `variable` correction,\n",
    "as we have no invariant character in our matrix. If we also lacked\n",
    "parsimony non-informative characters, we would use the coding `informative`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "phyMorpho ~ dnPhyloCTMC(tree=phylogeny, siteRates=rates_morpho, Q=Q_morpho, type=\"Standard\", coding=\"variable\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    ">Remember to change all filenames for the output, e.g., from `output/mk.log` to `output/mkv.log`.\n",
    "{:.instruction}\n",
    "\n",
    "That’s all you need to do! Now run this script in RevBayes.\n",
    "\n",
    "\n",
    "\n",
    "## Example: Relaxing the Assumption of Equal Transition Probabilities\n",
    "\n",
    ">Make a copy of the Rev script you made earlier. Call it\n",
    ">`mcmc_mk_dicretized.Rev`. This new script will\n",
    ">contain the new model parameters and models.\n",
    "\n",
    "\n",
    "<img src=\"figures/tikz/morpho_gm.png\" width=\"400\" /> \n",
    "_Graphical model demonstrating the\n",
    "discretized Beta distribution for allowing variable state frequencies._\n",
    "\n",
    "\n",
    "The Mk model makes a number of assumptions, but one that may strike you\n",
    "as unrealistic is the assumption that characters are equally likely to\n",
    "change from any one state to any other state. That means that a trait is\n",
    "as likely to be gained as lost. While this may hold true for some\n",
    "traits, we expect that it may be untrue for many others.\n",
    "\n",
    "RevBayes has functionality to allow us to relax this assumption. We do\n",
    "this by specifying a Beta prior on state frequencies. Remember from the\n",
    "CTMC lesson lesson that stationary frequencies impact how likely we are\n",
    "to see changes in a character. For example, it may be very likely, in a\n",
    "character, to change from 0 to 1. But if the frequency of 0 is very low,\n",
    "we will still seldom see this change.\n",
    "\n",
    "We can exploit the relationship between state frequencies and observed\n",
    "changes to allow for variable Q matrices across characters {% ref morpho_graphical_model %}. To\n",
    "do this, we generate a Beta distribution on state frequencies, and use\n",
    "the state frequencies from that Beta distribution to generate a series\n",
    "of Q-matrices to use to evaluate our data {% cite Pagel2004 %}.\n",
    "\n",
    "This type of model is called a **mixture model**. There are assumed to\n",
    "be subdivisions in the data, which may require different parameters (in\n",
    "this case, state frequencies). These subdivisions are not defined *a\n",
    "priori*. This model has previously been shown to be effective for a\n",
    "range of empirical and simulated datasets \\cite{Wright2016}.\n",
    "\n",
    "\n",
    "\n",
    "### Modifying the Rev-script %}\n",
    "\n",
    "At each place in which the output files are specified in the MCMC file,\n",
    "change the output path so you do not overwrite the output from the\n",
    "previous exercise. For example, you might call your output file\n",
    "`output/mk_discretized.log` and `output/mk_discretized.trees`. Change\n",
    "source statement to indicate the new model file.\n",
    "\n",
    "\n",
    "We will use a discretized Beta distribution to place a prior on state frequencies. \n",
    "The Beta distribution has two parameters, $\\alpha$ and $\\beta$. These two\n",
    "parameters specify the shape of the distribution. State frequencies will\n",
    "be evaluated according to this distribution, in the same way that rate\n",
    "variation is evaluated according to the Gamma distribution. The\n",
    "discretized distribution is split into multiple classes, each with it's\n",
    "own set of frequencies for the 0 and 1 characters. The number of classes\n",
    "can vary; we have chosen 4 for tractability. Note that we need to make sure that this\n",
    "discretization results into a symmetric model, therefore we will use only one parameter\n",
    "for the Beta distribution: `beta_scale` so that $\\alpha = \\beta$."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "    num_cats = 4\n",
    "    beta_scale ~ dnLognormal( 0.0, sd=2*0.587405 )\n",
    "    moves[mvi++] = mvScale(beta_scale, lambda=1, weight=5.0 )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Above, we initialized the number of categories, the parameters to the\n",
    "Beta distribution, and the moves on the parameters to the Beta.\n",
    "\n",
    "Next, we set the categories to each represent a quadrant of the Beta\n",
    "distribution specified by the `beta_scale`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "    cats := fnDiscretizeBeta(beta_scale, beta_scale, num_cats)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you were to print the `cats` variable, you would see a list of state\n",
    "frequencies like so:\n",
    "\n",
    "Using these state frequencies, we will generate a new vector of Q\n",
    "matrices. Because we are varying the state frequencies, we must use a Q\n",
    "matrix generation function that allows for state frequencies to vary as\n",
    "a parameter. We will, therefore, use the `fnF81` function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "    for (i in 1:cats.size())\n",
    "    {\n",
    "        Q[i] := fnF81(simplex(abs(1-cats[i]), cats[i]))\n",
    "    }"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Additionally, in RevBayes we need to specify the probablities that a site evolves according to one\n",
    "of the $Q$-matrices. For this model the probabilities must be equal because we need to guarantee that\n",
    "the model is symmetric. This, we use a `simplex` function to create a vector that sums to 1.0."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "    matrix_probs <- simplex( rep(1,num_cats) )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The only other specification that needs to change in the model specification is\n",
    "the CTMC:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "    phyMorpho ~ dnPhyloCTMC(tree=phylogeny, siteRates=rates_morpho, Q=Q, type=\"Standard\", coding=\"variable\", siteMatrices=matrix_probs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You will notice that we have added a command to tell the CTMC that we have\n",
    "multiple site matrices that will be applied to different characters in\n",
    "the matrix.\n",
    "\n",
    "\n",
    "### Set-Up the MCMC\n",
    "\n",
    "The MCMC chain set-up does not need to change. Run the new MCMC file,\n",
    "just as you ran the plain Mk file. This estimation will take longer than\n",
    "the Mk model, due to increased model complexity.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "### Evaluate and Summarize Your Results\n",
    "\n",
    "\n",
    "\n",
    "#### Evaluate MCMC | subsubsec_Eval_MCMC\n",
    "\n",
    "We will use `Tracer`to evaluate the MCMC samples from our\n",
    "three estimations. Load all three of the MCMC logs into the\n",
    "`Tracer` window. Highlight all three files in\n",
    "the upper left-hand viewer ({% ref add_files %}) by right- or\n",
    "command-clicking all three files.\n",
    "\n",
    "\n",
    "<img src=\"figures/AddFiles.png\" width=\"50%\"/> \n",
    "_Highlight all three files for model comparison._\n",
    "\n",
    "Once all three trace logs are loaded and highlighted, first look at the\n",
    "estimated likelihoods. You will notice that the Mk model, as\n",
    "originally proposed by \\cite{Lewis2001} is improved by allowing any state\n",
    "frequency heterogeneity. The discretized model and the Dirichlet\n",
    "model both represent improvements, but are fairly close in likelihood\n",
    "score to each other Tracer. Likely, we would need to\n",
    "perform stepping stone model assessment to truly tell if the more\n",
    "complicated model is statistically justified. \n",
    "\n",
    "\n",
    "<img src=\"figures/likelihoods.png\" width=\"100%\"/> \n",
    "_Comparison of likelihood scores for all three models._\n",
    "\n",
    "Click on the `Trace` panel. In the lower left hand corner, you will\n",
    "notice an option to color each trace by the file it came from. Choose\n",
    "this option (you may need to expand the window slightly to see it). Next\n",
    "to this option, you can also see an option to add a legend to your trace\n",
    "window. The results of this coloring can be seen in\n",
    "{% ref coltrace %}. When the coloring is working, you will see that\n",
    "the Mk model mixes quite well, but that mixing becomes worse as we relax\n",
    "the assumption of equal state frequencies. This is because we are\n",
    "greatly increasing model complexity. Therefore, we would need to run the\n",
    "MCMC chains longer if we were to use these analyses in a paper.\n",
    "\n",
    "<img src=\"figures/colortrace.png\" width=\"80%\"/> \n",
    "_The Trace window. The traces are colored by which version of the Mk model \n",
    "they correspond to._\n",
    "\n",
    "We are interested in two aspects of the posterior distribution. First,\n",
    "all analyses correct for the biased sampling of variable characters\n",
    "except for the simple analysis. Then, we expect the\n",
    "tree_length variable to be greater for simple\n",
    "than for the remaining analyses, because our data are enriched for\n",
    "variation. {% ref tracer_tree_length %} shows that\n",
    "`tree_length` is approximately 30% greater for\n",
    "simple than for `mk_simple`, which are\n",
    "identical except that `mk_simple` corrects for sampling\n",
    "bias. To compare these densities, click the “Marginal Prob Distribution”\n",
    "tab in the upper part of the window, highlight all of the loaded Trace\n",
    "Files, then select `tree_length` from the list of Traces.\n",
    "\n",
    "<img src=\"figures/results/tracer_tree_length.png\" width=\"50%\" /> \n",
    "_Posterior tree length estimates._\n",
    "\n",
    "Second, we are interested in characterizing the degree of heterogeneity\n",
    "estimated by the beta-discretized model. If the data were distributed by\n",
    "a single morphological rate matrix, then we would expect to see very\n",
    "little variation among the different values in cats, and\n",
    "very large values for the shape and scale parameters of the\n",
    "discrete-beta distribution. For example, if alpha_ofbeta =\n",
    "beta_ofbeta = 1000, then that would cause all discrete-beta\n",
    "categories to have values approaching 0.5, which approximates a\n",
    "symmetric Mk model.\n",
    "\n",
    "<img src=\"figures/results/cats.png\" width=\"50%\" /> \n",
    "_Posterior discretized state frequencies for the discrete-beta model._\n",
    "\n",
    "\n",
    "<img src=\"figures/results/alpha_beta.png\" width=\"50%\" /> \n",
    "_Posterior alpha and beta parameters for the discrete-beta model._\n",
    "\n",
    "The tracer output above shows that the four discrete-beta state\n",
    "frequencies do not all have the exact same value. In addition,\n",
    "the Tracer output of the alpha and beta parameters shows that the priors on the\n",
    "discrete-beta distribution are small enough that we expect to see\n",
    "variance among cat values. If the data contained no\n",
    "information regarding the distribution of cat values, then\n",
    "the posterior estimates for alpha_ofbeta and\n",
    "beta_ofbeta would resemble the prior.\n",
    "\n",
    "\n",
    "### Summarizing tree estimates\n",
    "\n",
    "The morphology trees estimated in Section Mk Model and\n",
    "Section Discretized Beta are summarized using a majority rule consensus tree\n",
    "(MRCT). Clades appearing in $p>0.5$ of posterior samples are resolved in\n",
    "the MRCT, while poorly support clades with $p \\leq 0.5$ are shown as\n",
    "unresolved polytomies. Poor phylogenetic resolution might be caused by\n",
    "having too few phylogenetically informative characters, or it might be\n",
    "due to conflicting signals for certain species relationships. Because\n",
    "phylogenetic information is generated through model choice, let's\n",
    "compare our topological estimates across models.\n",
    "\n",
    "<img src=\"figures/results/mk_discretized_majrule_tre.png\" width=\"75%\" /> \n",
    "_Majority rule consensus tree for the beta-discretized Mkv analysis._\n",
    "\n",
    "The MRCTs for the simple model with and without the +v correction are\n",
    "very similar to that for the discretized-beta model\n",
    "({% ref mk_discretized_majrule %}). Note that the scale bars for\n",
    "branch lengths differ greatly, indicating that tree length estimates are\n",
    "inflated without the +v correction, just as we saw when comparing the\n",
    "posterior tree length densities. In general, it is important to assess\n",
    "whether your results are sensitive to model assumptions, such as the\n",
    "degree of model complexity, and any mechanistic assumptions that\n",
    "motivate the model's design. In this case, our tree estimate appears to\n",
    "be robust to model complexity."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "RevBayes",
   "language": "bash",
   "name": "revbayes_kernel"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "r"
   },
   "file_extension": ".Rev",
   "help_links": [
    {
     "text": "RevBayes",
     "url": "https://revbayes.org"
    },
    {
     "text": "RevBayes Kernel",
     "url": "https://github.com/sdwfrost/revbayes_kernel"
    },
    {
     "text": "MetaKernel Magics",
     "url": "https://github.com/calysto/metakernel/blob/master/metakernel/magics/README.md"
    }
   ],
   "mimetype": "text/x-rsrc",
   "name": "RevBayes",
   "pygments_lexer": "R"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
